File: AED_2/Trabalho_pratico_2/Q09.java
import java.util.Date;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.FileNotFoundException;
import java.io.BufferedReader;
import java.io.FileReader;

class Jogador{
    private int id;
    private String nome;
    private int altura;
    private int peso;
    private String universidade;
    private int anoNascimento;
    private String cidadeNascimento;
    private String estadoNascimento;
    
    Jogador(){
        setAll(0, "", 0, 0, "", 0, "", "");
    }

    Jogador(int id){
        ler(id);
    }
    Jogador(int id, String nome, int altura, int peso, String universidade, int anoNascimento, String cidadeNascimento, String estadoNascimento){
        setAll(id, nome, altura, peso, universidade, anoNascimento, cidadeNascimento, estadoNascimento);
    }

    //metodos set
    public void setId(int id){
        this.id = id;
    }
    public void setNome(String nome){
        if(nome == null){
             this.nome = "nao informado";
         } else {
             this.nome = nome;
         }
    }
    public void setAltura(int altura){
        this.altura = altura;
    }
    public void setPeso(int peso){
        this.peso = peso;
    }
    public void setUniversidade(String universidade){
       if(universidade == ""){
            this.universidade = "nao informado";
        } else {
            this.universidade = universidade;
        }
    }
    
    public void setAnoNascimento(int ano){
        this.anoNascimento = ano;
    }
    public void setCidadeNascimento(String cidade){
        if(cidade == ""){
            this.cidadeNascimento = "nao informado";
        } else {
            this.cidadeNascimento = cidade;
        }
    }
    public void setEstadoNascimento(String estado){
        if(estado == ""){
            this.estadoNascimento = "nao informado";
        } else {
            this.estadoNascimento = estado;
        }
    }

    public void setAll(int id, String nome, int altura, int peso, String universidade, int anoNascimento, String cidadeNascimento, String estadoNascimento){
        setId(id);
        setNome(nome);
        setAltura(altura);
        setPeso(peso);
        setUniversidade(universidade);
        setAnoNascimento(anoNascimento);
        setCidadeNascimento(cidadeNascimento);
        setEstadoNascimento(estadoNascimento);

    }

    //metodos get
    public int getId(){
        return id;
    }
    public String getNome(){
        return nome;
    }
    public int getAltura(){
        return altura;
    }
    public int getPeso(){
        return peso;
    }
    public String getUniversidade(){
        return universidade;
    }
    public int getAnoNascimento(){
        return anoNascimento;
    }
    public String getCidadeNascimento(){
        return cidadeNascimento;
    }
    public String getEstadoNascimento(){
        return estadoNascimento;
    }

    //split para a leitura
    private String[] split(String linha){
        String[] splitted = {"", "", "0", "0", "", "0", "", ""};
        int pos = 0;
        if(linha != null){
            for(int i=0; i<linha.length(); i++){
                String aux = "";
 
                while(i<linha.length() && linha.charAt(i) != ','){
                    aux += linha.charAt(i++);
                }
 
                splitted[pos] = aux;
                pos++;
            }
        } 
        return splitted;
    }
    //metodo de leitura
    public void ler(int id) throws NumberFormatException{
        try{
            File file = new File("/tmp/players.csv");
            BufferedReader arq = new BufferedReader( new FileReader(file) );
            arq.readLine(); //pular primeira linha das informacoes das colunas
            String[] splitted = split(arq.readLine());
            while( Integer.valueOf(splitted[0]) != id )
                splitted = split(arq.readLine());
            
            String nome = splitted[1];
            int altura = Integer.valueOf(splitted[2]);
            int peso = Integer.valueOf(splitted[3]);
            String universidade = splitted[4];
            int anoNascimento = Integer.valueOf(splitted[5]);
            String cidadeNascimento = splitted[6];
            String estadoNascimento = splitted[7];

            setAll(id, nome, altura, peso, universidade, anoNascimento, cidadeNascimento, estadoNascimento); 
        } catch(IOException e){
            MyIO.println("Um erro ocorreu.");
            e.printStackTrace();
        }
    }
    
    //metodo de escrita
    public void imprimir(){
        MyIO.println("["+ getId() + " ## " + getNome() + " ## " + getAltura() +  " ## " + getPeso() + " ## " + getAnoNascimento() + " ## " + getUniversidade() + " ## " + getCidadeNascimento() + " ## " + getEstadoNascimento() + "]");
    }

    //metodo de clone
 
    public Jogador clone() {
        return new Jogador(id, nome, altura, peso, universidade, anoNascimento, cidadeNascimento, estadoNascimento);
    }
}



class Q09{
    private static int comp = 0, troca = 0;
    private static Jogador[] jogador;
    private static Jogador[] aumentarArray(int tam){
        Jogador[] temp = new Jogador[tam+1];
        for(int i=0; i < (tam); i++){
            temp[i] = jogador[i];
        }
        temp[tam] = new Jogador();
        return temp;
    }

    private static void swap(int i, int j){
        Jogador tmp = jogador[i];
        jogador[i] = jogador[j];
        jogador[j] = tmp;
        troca+=3;
    }
    private static boolean comp(Jogador A, Jogador B) {
        boolean compare = A.getAltura() > B.getAltura();
        if ( A.getAltura() == B.getAltura() ) {
            compare = A.getNome().compareTo(B.getNome()) > 0;
        }
        comp++;
        return compare;
    }
    private static void construir(int tamHeap){
        for (int i = tamHeap; i > 1 && comp(jogador[i], jogador[i/2]); i/=2){
            swap(i, i/2);
        } 
    }

    public static int getMaiorFilho(int i, int tamHeap){
        int filho;
        if (2*i == tamHeap || comp(jogador[2*i], jogador[2*i+1])){
            filho = 2*i;
            comp++;
        } else {
            filho = 2*i + 1;
            comp+=2;
        }
        return filho;
    }

    public static void reconstruir(int tamHeap){
        int i = 1;
        while(i <= (tamHeap/2)){
            int filho = getMaiorFilho(i, tamHeap);
            if(comp(jogador[filho], jogador[i])){
                swap(i, filho);
                i = filho;
                comp++;
            }else{
                i = tamHeap;
                comp+=2;
            }
        }
    }
    private static void heapSort(int tam){
        // Alterar o vetor ignorando a posicao zero
        Jogador[] tmp = new Jogador[tam + 1];
        for (int i = 0; i < tam; i++) {
            tmp[i + 1] = jogador[i];
        }
        jogador = tmp;
        //construindo o heap
        for (int tamHeap = 2; tamHeap <= tam ; tamHeap++){
            construir(tamHeap);
        }
        
        //ordenando
        int tamTmp = tam;
        while(tamTmp > 1){
            swap(1,tamTmp--);
            reconstruir(tamTmp);
        }

        // Alterar o vetor para voltar a posicao zero
        tmp = jogador;
        jogador = new Jogador[tam];
        for (int i = 0; i < tam; i++) {
            jogador[i] = tmp[i + 1];
        }
    }
    
    public static void main(String[] args){
        try{
            int tam = 0;
            jogador = new Jogador[0];
            String id;
            String nomeLido;
    
            //ler id's e armazenar quantidade lida (tam) e Jogadores correspondentes
            id = MyIO.readLine();
            while(!id.equals("FIM")){
                jogador = aumentarArray(tam);
                jogador[tam++].ler(Integer.valueOf(id));
                id = MyIO.readLine();
            }
            
            long inicio = new Date().getTime();
            heapSort(tam);

            for(int i=0; i<tam; i++){
                jogador[i].imprimir();
            }
            long fim = new Date().getTime();
            File arq = new File("matricula_heapsort.txt");


            FileWriter fw = new FileWriter(arq);
            fw.write("805413\t" + "\t" + comp+ "\t" + troca + "\t" + (fim-inicio)/1000.0);
            fw.close();
        } catch (IOException e){
            e.printStackTrace();
        }catch(NumberFormatException e){
            MyIO.println("Id invalido");
        }

    }
}

File: Algoritmos_e_Estrutura_de_Dados_II/Trabalho_pratico_4/q09.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
    
#define TAM 60
#define DIR "/tmp/players.csv"

struct Jogador{
    int id;
    char nome[TAM];
    int altura;
    int peso;
    char universidade[TAM];
    int anoNascimento;
    char cidadeNascimento[TAM];
    char estadoNascimento[TAM];
};
typedef struct Jogador Jogador;

//--------------------Funcoes-Jogador------------------//
  //----funções set---//
  void setId(int id, Jogador* jogadorLido){
      jogadorLido->id = id;
  }
  void setNome(char nome[], Jogador* jogadorLido){
      if(nome[0]=='\0'){
          strcpy(jogadorLido->nome, "nao informado");
      } else {
          strcpy(jogadorLido->nome, nome);
      }
  }
  void setAltura(int altura, Jogador* jogadorLido){
      jogadorLido->altura = altura;
  }
  void setPeso(int peso, Jogador* jogadorLido){
      jogadorLido->peso = peso;
  }
  void setUniversidade(char universidade[], Jogador* jogadorLido){
      if(universidade[0] == '\0'){
             strcpy(jogadorLido->universidade, "nao informado");
      } else {
             strcpy(jogadorLido->universidade, universidade);
      }
  }
  void setAnoNascimento(int anoNascimento, Jogador* jogadorLido){
      jogadorLido->anoNascimento = anoNascimento;
  }
  void setCidadeNascimento(char cidadeNascimento[], Jogador* jogadorLido){
      if(cidadeNascimento[0] == '\0'){
          strcpy(jogadorLido->cidadeNascimento, "nao informado");
      } else {
          strcpy(jogadorLido->cidadeNascimento, cidadeNascimento);
      }
  }
  void setEstadoNascimento(char estadoNascimento[], Jogador* jogadorLido){
      if(estadoNascimento[0] == '\0'){
          strcpy(jogadorLido->estadoNascimento, "nao informado");
      } else {
          strcpy(jogadorLido->estadoNascimento, estadoNascimento);
      }
  }
  void setAll(int id, char nome[], int altura, int peso, char universidade[], int anoNascimento, char cidadeNascimento[], char estadoNascimento[], Jogador* jogadorLido){
      setId(id, jogadorLido);
      setNome(nome, jogadorLido);
      setAltura(altura, jogadorLido);
      setPeso(peso, jogadorLido);
      setUniversidade(universidade, jogadorLido);
      setAnoNascimento(anoNascimento, jogadorLido);
      setCidadeNascimento(cidadeNascimento, jogadorLido);
      setEstadoNascimento(estadoNascimento, jogadorLido);
  }

  //----funcao clone----//
  Jogador* clone(Jogador *j){
      Jogador* jClone = (Jogador*) malloc(sizeof(Jogador));
      setAll(j->id, j->nome, j->altura, j->peso, j->universidade, j->anoNascimento, j->cidadeNascimento, j->estadoNascimento, jClone);
      return jClone;
  }

  //----funcao split----//
  void splitJogador (char valLido[], char valFinal[], int virgulasPuladas){
      char aux;
      int posL = 0, posF = 0;
      while(virgulasPuladas>0){
          if(valLido[posL++]==',')
              virgulasPuladas--;
      }

      aux = valLido[posL++];
      while( aux != ',' && aux != '\n' ) {
          valFinal[posF++] = aux;
          aux = valLido[posL++];
      }
      valFinal[posF] = '\0';
  }


  //----funcao de leitura----//
  Jogador* lerJogador(int id){
      Jogador* jogador = NULL;

      FILE* arq = fopen(DIR, "r");
      if(arq == NULL)
          printf("Arquivo nao aberto");
      else{
          char idChar[TAM];
          char nome[TAM];
          char altura[TAM];
          char peso[TAM];
          char universidade[TAM];
          char nascimento[TAM];
          char cidade[TAM];
          char estado[TAM];
          char line[8*TAM];

          do{
              int tam = 0;
              char c;
              do{
                  c = fgetc(arq);
                  line[tam++] = c;
              }while ( c != '\n'  && c != EOF);
              splitJogador(line, idChar, 0);
          } while (atoi(idChar) != id && line[0] != EOF);

          if(line[0] == EOF){
              printf("Id inexistente");
          } else {
              jogador = (Jogador*) malloc(sizeof(Jogador));

              setId(atoi(idChar), jogador);

              splitJogador(line, nome, 1);
              setNome(nome, jogador);

              splitJogador(line, altura, 2);
              setAltura(atoi(altura), jogador);

              splitJogador(line, peso, 3);
              setPeso(atoi(peso), jogador);

              splitJogador(line, universidade, 4);
              setUniversidade(universidade, jogador);

              splitJogador(line, nascimento, 5);
              setAnoNascimento(atoi(nascimento), jogador);

              splitJogador(line, cidade, 6);
              setCidadeNascimento(cidade, jogador);

              splitJogador(line, estado, 7);
              setEstadoNascimento(estado, jogador);
          }
      }
      fclose(arq);
      return jogador;
 }

//-----------------Fim-funcoes-Jogador----------------/

struct No{
    struct No *prox;
    Jogador *thisJ;
};
typedef struct No No;

//----------------Funcoes-No------------------------//

No* newNo(Jogador* add){
    No* new = (No*) malloc(sizeof(No));
    new->thisJ = add;
    return new;
}

//---------------Fim-funcoes-No-------------------//

struct Lista{
    No* primeiro;
    No* ultimo;
};
typedef struct Lista Lista;

//--------------------Funcoes-Lista-----------------//

Lista* newLista(){
    Lista *lista = (Lista*) malloc(sizeof(Lista));
    lista->primeiro = lista->ultimo = (No*) malloc(sizeof(No));
    return lista;
}

void insertLista(Jogador* add, Lista* lista){
    lista->ultimo->prox = newNo(add);
    lista->ultimo = lista->ultimo->prox;
}

bool searchLista(char *nome, Lista *lista){
    bool resultado = false;
    No* i = lista->primeiro->prox;
    while(i != lista->ultimo){
        if(strcmp(nome, i->thisJ->nome) == 0){
            resultado = true;
            i = lista->ultimo;
        } else {
            i = i->prox;
        }
    }
    if(strcmp(nome, i->thisJ->nome) == 0)
            resultado = true;

    return resultado;
}

//----------------Fim-funcoes-Lista---------------//

struct Hash{
    Lista **vetor;
}; 
typedef struct Hash Hash;

//------------------funcoes-Hash---------------------//
static int tamTab = 25;

Hash* newHash(){
    Hash *hash = (Hash*) malloc(sizeof(Hash));
    hash->vetor = (Lista**) malloc(tamTab*sizeof(Lista*));
    return hash;
}

void insertHash(Jogador *add, Hash *hash){
    if(hash->vetor[add->altura % tamTab] == NULL)
        hash->vetor[add->altura % tamTab] = newLista();
    insertLista(add, hash->vetor[add->altura % tamTab]);
}

bool searchHash(char* nome, Hash *hash){
    bool resultado = false;
    for(int i=0; i<tamTab; i++){
        if(hash->vetor[i] != NULL){
            //printf("%b\n", searchLista(nome, hash->vetor[i]));
            if(searchLista(nome, hash->vetor[i])){
                resultado = true;
                i = tamTab;
            }
        }
    }
    return resultado;
}


int main() {
    Hash* hash = newHash();
    char in[50];
    char charProblematico[2] = {'\r'};

    scanf("%[^\n]\n", in);
    in[strcspn(in, charProblematico)] = '\0';
    while(strcmp(in, "FIM")){
        Jogador* tmp = lerJogador(atoi(in));
        if(tmp != NULL){
            insertHash(tmp, hash);
        } else {
	    	printf("Erro: id invalido");
		}
        scanf("%[^\n]\n", in);
        in[strcspn(in, charProblematico)] = '\0';
    }
   
    scanf("%[^\n]\n", in);
    in[strcspn(in, charProblematico)] = '\0';
    while(strcmp(in, "FIM") && in[0] != 4 ){ 
        printf("%s ", in);
        if(searchHash(in, hash)){
            printf("SIM\n");
        } else {
            printf("NAO\n");
        }
        scanf("%[^\n]\n", in);
        in[strcspn(in, charProblematico)] = '\0';
    }

    return 0;
}
